docs/network-layer.md

```diff
@@ -0,0 +1,70 @@
+
+1. The client versions are compatible
+2. The wire versions are compatible
+3. The peers have *compatible* genesis blocks. *Compatible* in this context means that one of the other must provide a set of *genesis blocks* which should be a subset of others.
+
+When the *handshakes* has been successfully completed then the peers will *promote* each other into their corresponding lists of peers, i.e. the peers are now connected and can communicate with each other, hence the peers can now relay messages between them, carry out *consensus* and *catchup* etc.
+
+## Control Flow
+When the node is running, it continously checks for incoming connections and polls for events related to the already established connections. Furthermore the node performs several housekeeping tasks wrt. networking related matters. More specifically the node checks that its current connections are *healthy* i.e., verifying reasonable latency, removing connections of which a handshake is missing etc. If any connections are deemed *unhealthy* then the node will close the connection and remove the peer from its set of peers. Note that depending on the level of *unhealthy* the node will perhaps *soft ban* a peer (more on this below).
+
+
+If the node has exceeded its maximum amount of peers it will remove a random subset of its peers. Note that the node will never remove any of the *pinned peers*. 
+The node then proceeds to check if it should unban any of the *soft banned* peers and does so if that is the case. Note that *soft bans* has a duration of 5 minutes.
+
+Afterwards the node may try to *re-bootstrap* again (if the node did not recently *bootstrap*). Finally if the node still lacks peers then it will send out *GetPeers* messages to its peers.
+
+
+The node listens on its established connections and processes the received messages accordingly. If the message was *re-broadcastable*, then the node will relay the message to its own peers such that the message will propagate throughout the network. Before processing the message the node asserts that it has not yet seen the message before or otherwise the node will not process the *duplicated* message. 
+
+The control flow of the *network layer* can be summarized as follows:
+
+1. Listen for incoming connections
+2. Process network messages
+3. (Optionally) Carry out housekeeping tasks if a suffcient time period has passed since last time a housekeeping was done.
+4. (Optionally) Establish new outgoing connections if required
+5. Repeat
+
+## Processing Network Message
+When the node receives a network message on one of its connections it will either dispatch the network message to the *consensus layer* or immediately handle the network message if it was only *network related*. An example of the former could be that the node was requested for peers by the *GetPeers* message, responding to a *Ping*, *Join*-/*Leave* network messages etc. 
```

Comment:
```suggestion
When the node receives a network message on one of its connections it will either dispatch the network message to the *consensus layer* or immediately handle the network message if it was only *network related*. An example of the latter could be that the node was asked for peers by the *GetPeers* message, responding to a *Ping*, *Join*-/*Leave* network messages etc. 
```